File: slp/backend-dotnet/Features/Progress/ProgressService.cs
using System.Text.Json;

namespace backend_dotnet.Features.Progress;

public interface IProgressService
{
    Task<ProgressDto?> GetProgressAsync(int userId, int sourceId);
    Task<ProgressDto> UpdateProgressAsync(int userId, int sourceId, UpdateProgressRequest request);
}

public class ProgressService : IProgressService
{
    private readonly IProgressRepository _repo;

    public ProgressService(IProgressRepository repo)
    {
        _repo = repo;
    }

    public async Task<ProgressDto?> GetProgressAsync(int userId, int sourceId)
    {
        var entity = await _repo.GetAsync(userId, sourceId);
        if (entity == null)
            return new ProgressDto { SourceId = sourceId, LastPosition = null, UpdatedAt = DateTime.UtcNow };

        return MapToDto(entity);
    }

    public async Task<ProgressDto> UpdateProgressAsync(int userId, int sourceId, UpdateProgressRequest request)
    {
        var json = JsonSerializer.Serialize(request.LastPosition);
        await _repo.UpsertAsync(userId, sourceId, json);
        var entity = await _repo.GetAsync(userId, sourceId);
        return MapToDto(entity!);
    }

    private static ProgressDto MapToDto(UserSourceProgress p)
    {
        object? lastPos = null;
        if (!string.IsNullOrEmpty(p.LastPosition))
        {
            try { lastPos = JsonSerializer.Deserialize<object>(p.LastPosition); } catch { }
        }
        return new ProgressDto
        {
            SourceId = p.SourceId,
            LastPosition = lastPos,
            UpdatedAt = p.UpdatedAt
        };
    }
}

File: slp/backend-nestjs/src/modules/progress/progress.service.ts
import { Injectable } from '@nestjs/common';
import { ProgressRepository } from './progress.repository';
import { SourceRepository } from '../source/source.repository';
import { ProgressDto } from './dto/progress.dto';
import { UpdateProgressRequest } from './dto/update-progress-request.dto';
import { UserSourceProgress } from './progress.entity';

@Injectable()
export class ProgressService {
  constructor(
    private readonly progressRepo: ProgressRepository,
    private readonly sourceRepo: SourceRepository,
  ) {}

  private mapToDto(entity: UserSourceProgress): ProgressDto {
    return {
      sourceId: entity.sourceId,
      lastPosition: entity.lastPosition,
      updatedAt: entity.updatedAt,
    };
  }

  async getProgress(userId: number, sourceId: number): Promise<ProgressDto> {
    const entity = await this.progressRepo.getProgress(userId, sourceId);
    if (!entity) {
      // Return a default progress with null position
      return {
        sourceId,
        lastPosition: null,
        updatedAt: new Date(),
      };
    }
    return this.mapToDto(entity);
  }

  async updateProgress(
    userId: number,
    sourceId: number,
    request: UpdateProgressRequest,
  ): Promise<ProgressDto> {
    // Verify that the source exists and belongs to the user
    const source = await this.sourceRepo.getById(sourceId);
    if (!source || source.userId !== userId) {
      throw new Error('Source not found or does not belong to user');
    }

    const updated = await this.progressRepo.upsertProgress(
      userId,
      sourceId,
      request.lastPosition,
    );
    return this.mapToDto(updated);
  }
}